internal/gitinterface/sync.go

```diff
@@ -0,0 +1,147 @@
+	}
+
+	pushOpts := &git.PushOptions{
+		RemoteName: remoteName,
+		RefSpecs:   refs,
+		Atomic:     true,
+	}
+
+	err = remote.PushContext(ctx, pushOpts)
+	if errors.Is(err, git.NoErrAlreadyUpToDate) {
+		return nil
+	}
+	return err
+}
+
+// Push constructs refspecs for the specified Git refs and pushes from the repo
+// to the remote. For more information on the Git refspec, please consult:
+// https://git-scm.com/book/en/v2/Git-Internals-The-Refspec.
+//
+// The refspecs are constructed to be fast-forward only.
+func Push(ctx context.Context, repo *git.Repository, remoteName string, refs []string) error {
+	refSpecs := make([]config.RefSpec, 0, len(refs))
+	for _, r := range refs {
+		refSpec, err := RefSpec(repo, r, "", true)
+		if err != nil {
+			return err
+		}
+		refSpecs = append(refSpecs, refSpec)
+	}
+
+	return PushRefSpec(ctx, repo, remoteName, refSpecs)
+}
+
+// FetchRefSpec fetches to the repo from the specified remote using
+// pre-constructed refspecs. For more information on the Git refspec, please
+// consult: https://git-scm.com/book/en/v2/Git-Internals-The-Refspec.
+func FetchRefSpec(ctx context.Context, repo *git.Repository, remoteName string, refs []config.RefSpec) error {
+	remote, err := repo.Remote(remoteName)
+	if err != nil {
+		return err
+	}
+
+	fetchOpts := &git.FetchOptions{
+		RemoteName: remoteName,
+		RefSpecs:   refs,
+	}
+
+	err = remote.FetchContext(ctx, fetchOpts)
+	if errors.Is(err, transport.ErrEmptyRemoteRepository) || errors.Is(err, git.NoErrAlreadyUpToDate) {
+		return nil
+	}
+	return err
+}
+
+// Fetch constructs refspecs for the refs and fetches to the repo from the
+// specified remote. For more information on the Git refspec, please consult:
+// https://git-scm.com/book/en/v2/Git-Internals-The-Refspec.
+//
+// The refspecs are constructed to allow non-fast-forward fetches. Additionally,
+// trackRemote controls if the change is fetched to a local remote tracker.
+func Fetch(ctx context.Context, repo *git.Repository, remoteName string, refs []string, trackRemote bool) error {
+	refSpecs := make([]config.RefSpec, 0, len(refs))
+	for _, r := range refs {
+		var (
+			refSpec config.RefSpec
+			err     error
+		)
+		if trackRemote {
+			refSpec, err = RefSpec(repo, r, remoteName, false)
+			if err != nil {
+				return err
+			}
+		} else {
+			refSpec, err = RefSpec(repo, r, "", false)
+			if err != nil {
+				return err
+			}
+		}
+		refSpecs = append(refSpecs, refSpec)
+	}
+
+	return FetchRefSpec(ctx, repo, remoteName, refSpecs)
+}
+
+// CloneAndFetch clones a repository using the specified URL and additionally
+// fetches the specified refs. If trackRemote is set, the refs are fetched to a
+// local remote tracker.
+func CloneAndFetch(ctx context.Context, remoteURL, dir, initialBranch string, refs []string, trackRemote bool) (*git.Repository, error) {
+	repo, err := git.PlainCloneContext(ctx, dir, false, &git.CloneOptions{URL: remoteURL, ReferenceName: plumbing.ReferenceName(initialBranch)})
```

Comment:
I'd recommend avoiding the initial branch name if you can, or use HEAD. It'll save headache down the road when you deal with `main` vs `master` branches.

Reply:
I'll test this some more. My thought was `initialBranch` can be set to "" defaulting to whatever is the main branch on the remote.

Reply:
Done! Added a test to both Clone methods for when initialBranch is an empty string. When creating clone options, I'm explicitly checking length before setting it, I can confirm go-git defaults to HEAD when the initial branch is not set.

Reply:
This was only to simulate `git clone -b` so I think we're fine now.